src/plugins/visualizations/public/utils/saved_visualize_utils.ts

```
@@ -0,0 +1,362 @@
+        };
+      } else {
+        return mapHitSource(visTypes, savedObject);
+      }
+    }),
+  };
+}
+
+export async function getSavedVisualization(
+  services: {
+    savedObjectsClient: SavedObjectsClientContract;
+    search: any;
+    dataViews: any;
+  },
+  opts?: any
+) {
+  if (typeof opts !== 'object') {
+    opts = { id: opts };
+  }
+
+  const id = (opts.id as string) || '';
+  const savedObject = {
+    id,
+    displayName: SAVED_VIS_TYPE,
+    getEsType: () => SAVED_VIS_TYPE,
+    getDisplayName: () => SAVED_VIS_TYPE,
+    searchSource: opts.searchSource ? services.search.searchSource.createEmpty() : undefined,
+  } as { [key: string]: any };
+
+  const defaultsProps = getDefaults(opts);
+
+  if (!id) {
+    assign(savedObject, defaultsProps);
+    return Promise.resolve(savedObject);
+  }
+
+  const resp = await services.savedObjectsClient.get<Record<string, unknown>>(SAVED_VIS_TYPE, id);
+
+  if (!resp._version) {
+    throw new SavedObjectNotFound(SAVED_VIS_TYPE, id || '');
+  }
+
+  savedObject._source = cloneDeep(resp.attributes);
+
+  if (savedObject._source.visState) {
+    savedObject._source.visState = JSON.parse(savedObject._source.visState);
+  }
+
+  // assign the defaults to the response
+  defaults(savedObject._source, defaultsProps);
+
+  assign(savedObject, savedObject._source);
+  savedObject.lastSavedTitle = savedObject.title;
+
+  const meta = savedObject._source.kibanaSavedObjectMeta || {};
+
+  if (meta.searchSourceJSON) {
+    try {
+      let searchSourceValues = parseSearchSourceJSON(meta.searchSourceJSON);
+
+      if (opts.searchSource) {
+        searchSourceValues = injectSearchSourceReferences(
+          searchSourceValues as any,
+          resp.references
+        );
+        savedObject.searchSource = await services.search.searchSource.create(searchSourceValues);
+      } else {
+        savedObject.searchSourceFields = searchSourceValues;
+      }
+    } catch (error: any) {
+      if (
```

@VladLasitsa this stuff is only necessary for the legacy import - we need to create a valid saved vis instance which captures the error state. As this won't be used for legacy import, we can just bubble up the error.